File: 프로그래머스/2/42586. 기능개발/기능개발.cpp
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> progresses, vector<int> speeds) {
    vector<int> answer;
    int n = progresses.size(); // 기능의 개수
    int current = 0;           // 현재 배포할 기능의 인덱스

    while (current < n) {
        // 각 기능의 진행 상황 업데이트
        for (int i = 0; i < n; i++) {
            progresses[i] += speeds[i];
        }

        // 이번에 배포할 수 있는 기능들의 수를 셉니다.
        int count = 0;
        while (current < n && progresses[current] >= 100) {
            count++;
            current++; // 배포된 기능의 인덱스 이동
        }

        // 배포할 기능이 있다면 그 수를 저장
        if (count > 0) {
            answer.push_back(count);
        }
    }

    return answer;
}

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        ArrayList<Integer> list = new ArrayList<>();
        
        int cnt = 0;
        while(cnt < progresses.length){
            ArrayDeque<Integer> que = new ArrayDeque<>();
            
            int k = 0;
            for(int i=cnt; i<progresses.length; i++){
                progresses[i] += speeds[i];
                que.addLast(progresses[i]);
            }
            
            while(!que.isEmpty() && que.peekFirst() >= 100){
                que.removeFirst();
                cnt++;
                k++;
            }
            
            if(k>0) list.add(k);
        
        }
        
        int[] answer = new int[list.size()];
        for(int i=0; i<list.size(); i++){
            answer[i] = list.get(i);
        }
        
        
        return answer;
    }
}